### ISSUE TITLE
MongoDB IndexOptions type is missing 

### ISSUE THREAD

**Issue by rlawisch** [Automattic/mongoose#10590]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior? If the current behavior is a bug, please provide the steps to reproduce.**
In `Schema.index` second parameter, `{ unique: true }` was marked as invalid because `unique` is not a member of Mongoose's `IndexOption` interface.

```typescript
import { model, Schema } from "mongoose";

export interface User {
  email?: string;
  password?: string;
  createdAt?: Date;
  updatedAt?: Date;
}

const UserSchema = new Schema<User>(
  {
    email: { type: String, required: true },
    password: { type: String, required: true },
  },
  { timestamps: true }
);

UserSchema.index({ email: 1, campaign: 1 }, { unique: true });

export default model<User>("User", UserSchema);
```

EDIT: [See it on StackBlitz](https://stackblitz.com/edit/typescript-ceyd7z?file=user.model.ts)

The Mongoose's interface extends MongoDB's `IndexOptions` interface, which was not available in `@types/mongodb` version `3.6.20` resolved as a Mongoose dependency by NPM.
Manually installing an updated package (version `4.0.7`) fixed the error.

**What is the expected behavior?**
Using `unique` and other index options should be TS-compilable.

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**

- Node.js: `14.17.3`
- Mongoose: `5.13.7`
- MongoDB: `4.1.0`
- TypeScript: `4.3.5`

**Comment by hasezoey** [comment]
>  Manually installing an updated package (version 4.0.7) fixed the error.

are you sure about that? because since 4.x mongodb provides official types, making the `@types/mongodb` a stub package

**Comment by rlawisch** [comment]
Yes, I am. A few seconds after installing it, the error was gone. Also, the VSCode navigation breadcrumbs for the (fixed) interface is `node_modules > mongoose > node_modules > @types > mongodb > index.d.ts > IndexOptions`, which indicates it is still getting the interface definition from `@types/mongodb`.
I see MongoDB is providing official types, but I couldn't find `IndexOptions` among them.

**Comment by hasezoey** [comment]
>  I see MongoDB is providing official types, but I couldn't find IndexOptions among them.

mongodb 4.x is only supported from / since mongoose 6.0
btw, in mongodb's official types its `CreateIndexesOptions` (which is already changed to in mongoose 6.0)

> which indicates it is still getting the interface definition from @types/mongodb.

so what version of `@types/mongodb` is installed now?

**Comment by vkarpov15** [comment]
Since the error is gone, I'll close this for now.

**Comment by rlawisch** [comment]
@hasezoey 

> so what version of @types/mongodb is installed now?

Version `4.0.7` is installed and has the `IndexOptions` interface (`3.6.20` didn't).

**Comment by rlawisch** [comment]
@vkarpov15 It's gone, but was manually fixed. Shouldn't it be fixed for everyone else?
[Here's a StackBlitz proof/example](https://stackblitz.com/edit/typescript-ceyd7z?file=user.model.ts)

**Comment by hasezoey** [comment]
@rlawisch in your provided example, if you try to `Go to Definition` (to open mongoose's `index.d.ts`) on `UserSchema.index` and then go to the top of the page, you can see the error `Cannot find module 'mongodb' or its corresponding type declarations.(2307)`, which means that `stackblitz` is not properly installing (type) packages

**Comment by rlawisch** [comment]
@hasezoey yes, my bad, haven't seen this.
I've set up a test project, tried installing/uninstalling different versions, and just intermittently replicated the issue. As I could not replicate it with `5.13.8`, I guess @vkarpov15 can close it.
Thanks for your quick response, guys!

**Comment by tehmas** [comment]
@hasezoey

I'm facing this issue on mongoose `5.13.14` as well. When I use `SomeSchema.index({ fieldName: 1 }, { unique: true })`, I get error:
```
Argument of type '{ unique: boolean; }' is not assignable to parameter of type 'IndexOptions'.
  Object literal may only specify known properties, and 'unique' does not exist in type 'IndexOptions'.
```
When I go to definition of `.index()`, I came across this:
```
index(fields: IndexDefinition, options?: IndexOptions): this;
```
When go to definition of `IndexOptions`:
```
  interface IndexOptions extends mongodb.IndexOptions {
    expires?: number | string;
    weights?: AnyObject;
  }
```

I get the message "No definition found for IndexOptions" when I click "Go to definition" for `.IndexOptions` of `mongodb`.

`mongodb.d.ts` exists in my case.

**Comment by hasezoey** [comment]
> By an earlier version do you mean earlier version of mongoose?

no i by an earlier version i mean of `@types/mongodb`, for that you may also need to manually modify the lockfile and or use a overwrite when possible

**Comment by hasezoey** [comment]
@tehmas could you maybe provide output of `yarn why @types/mongodb` or `npm ls @types/mongodb`

**Comment by tehmas** [comment]
@hasezoey  out of `npm ls @types/mongodb`

```
└─┬ mongoose@5.13.14
  └── @types/mongodb@3.6.20
```

**Comment by hasezoey** [comment]
then this is weird, because i cannot get it to re-produce that error, because `@types/mongodb` is properly installed, `IndexOptions` should exist, maybe try some of the following:
- remove your `node_modules` and re-install
- upgrade typescript version if possible

**Comment by tehmas** [comment]
@hasezoey Tried both approaches (upgraded typescript to `4.7.4`) but it didn't resolve the issue.

I'm still getting the same error which @rlawisch has shared in stackblitz

**Comment by hasezoey** [comment]
i can see the error in stackblitz (even after upgrading to latest 5.13), but i cannot see the reason why and i cannot reproduce it locally

**Comment by titivermeesch** [comment]
My fix was to put this in my package.json:
```json
"resolutions": {
    "@types/mongodb": "3.6.20"
  }
```

Figured out that some other package was also depending on v4 types, which were apparently not compatible with my mongoose version v5.9

**Comment by vkarpov15** [comment]
I see the StackBlitz IDE showing a warning, but the project compiles fine, both locally and on StackBlitz. The following screenshot shows that I've made some edits and StackBlitz accepts them.

![image](https://user-images.githubusercontent.com/1620265/220160593-cbdd5016-f490-427c-99b8-d3fbdc8ed26b.png)

I'm going to chalk this up as a quirk of StackBlitz's IDE, since it looks like this project compiles fine on StackBlitz.

**Comment by doanthai** [comment]
In my case, my project run in monorepo with pnpm and everything is ok, but when put only a service to docker and install  with pnpm so this problem occur. I change package manage by yarn in Dockerfile and it work.

**Comment by hasezoey** [comment]
@doanthai this is quite a old issue and likely relevant information has changed, please open a new issue (maybe reference this one if it is similar)
